// builder/index.js
'use strict';

var $ = require('elements'),
	forOwn = require('mout/object/forOwn'),
	indexOf = require('mout/array/indexOf'),
	blocks = require('./blocks'),
	templates = require('./templates');

var Builder = function(){
	if (!(this instanceof Builder)) return new Builder();
	this.blockIds = {};
	this.blockId = 0;
	this.build();
	this.createBlocks();
	this.setEvents();
};

/**
 *
 */
Builder.prototype.build = function(){
	this.wrap = document.createElement('div');
	this.wrap.classList.add('builder');
	this.wrap.innerHTML = '<div class="builder--template">' +
			'<div class="builder--empty-layout">' +
				'Choose a template to start building' +
			'</div>' +
		'</div>' +
		'<div class="builder--block-types">' +
			'<ul></ul>' +
		'</div>';

	this.blockWrapper = this.wrap.querySelector('.builder--block-types ul');
	this.template = this.wrap.querySelector('.builder--template');

	this.placeholder = document.createElement('div');
	this.placeholder.classList.add('builder--placeholder');
};

/**
 *
 */
Builder.prototype.createBlocks = function(){
	var sorted = [];
	forOwn(blocks, function(block, key){
		block.meta.key = key;
		sorted.push(block);
	});
	sorted.sort(function(a, b){
		if (a.meta.name > b.meta.name) return 1;
		else if (a.meta.name < b.meta.name) return -1;
		return 0;
	});

	var i, li;
	for (i = 0; i < sorted.length; i++){
		li = document.createElement('li');
		li.setAttribute('data-block-type', sorted[i].meta.key);
		li.setAttribute('draggable', true);
		li.innerHTML = '<span class="block-name">' + sorted[i].meta.name + '</span>' +
			'<span class="block-description">' + sorted[i].meta.description + '</span>';
		this.blockWrapper.appendChild(li);
	}
};

Builder.prototype.cacheZones = function(){
	this.zones = {};
	var zones = this.template.querySelectorAll('.zone'),
		i, len = zones.length, zone;

	for (i = 0; i < len; i++){
		zone = zones[i].getAttribute('data-zone');
		if (!zone) continue;

		this.zones[zone] = zones[i];
	}
};

/**
 * Set all required drag events
 */
Builder.prototype.setEvents = function(){
	var self = this, isHeader;

	// drag block
	this.blockWrapper.addEventListener('dragstart', function(e){
		var target = e.target || e.srcElement,
			blockType = target.getAttribute('data-block-type');

		if (!blockType) return;

		self.dragging = true;
		self.isNewBlock = true;
		e.dataTransfer.dropEffect = 'copy';
		e.dataTransfer.setData('text', target.getAttribute('data-block-type'));
	});

	// end drag block
	document.body.addEventListener('dragend', function(e){
		self.dragEnd();
	});

	// indicate that the zones are droppable
	this.template.addEventListener('dragover', function(e){
		e.preventDefault(); // required for anything to be droppable at all
	});

	// check the entire body for dragovers otherwise sometimes leaving a zone is not detected
	document.body.addEventListener('dragover', function(e){
		var zone = self.getZone(e.target || e.srcElement);
		if (!zone){
			if (self.targetZone){
				self.zoneLeave(self.targetZone);
				self.targetZone = undefined;
			}
			return;
		}

		var zoneName = zone.getAttribute('data-zone');
		if (zoneName != self.targetZone){
			self.zoneLeave(self.targetZone);
			self.zoneEnter(zoneName);
			self.targetZone = zoneName;
		}
	});

	// enter zone
	this.template.addEventListener('dragenter', function(e){
		e.preventDefault();

		var target = e.target || e.srcElement,
			zone = self.getZone(target);

		if (!zone) return;

		var isPlaceholder = target == self.placeholder;
		if (isPlaceholder) return;

		var block = target;

		if (!block.classList.contains('.block')){
			block = $(target).parent('.block');
			if (!block) return;

			block = block[0];
		}

		var placeholderIndex = indexOf(zone.childNodes, self.placeholder),
			blockIndex = indexOf(zone.childNodes, block),
			method = 'after';

		if (placeholderIndex > blockIndex){
			method = 'before';
		}
		$(self.placeholder)[method](block);
	});

	this.template.addEventListener('mousedown', function(e){
		var target = e.target || e.srcElement;
		if (target.classList.contains('block--header')){
			isHeader = true;
		}
	});

	// dragging existing blocks
	this.template.addEventListener('dragstart', function(e){
		var target = e.target || e.srcElement;
		if (!isHeader){
			e.preventDefault();
			return;
		}
		e.dataTransfer.setData('text', ''); // needed for drag to work in FF
		isHeader = false;
		self.dragBlock = target;

		// this delay is needed cause the dragImage won't be created properly otherwise
		setTimeout(function(){
			var rect = self.dragBlock.getBoundingClientRect();
			self.placeholder.style.height = (rect.bottom - rect.top) + 'px';
			self.dragBlock.parentNode.insertBefore(self.placeholder, self.dragBlock);
			self.dragBlock.style.display = 'none';
		}, 1);
	});

	// drop zone
	this.template.addEventListener('drop', function(e){
		e.preventDefault();

		var zone = self.getZone(e.target || e.srcElement);
		if (!zone) return;

		var zoneName = zone.getAttribute('data-zone');

		if (self.isNewBlock){
			self.addBlock(zoneName, e.dataTransfer.getData('text'));
		} else {
			zone.insertBefore(self.dragBlock, self.placeholder);
			self.placeholder.style.height = '';
			zone.removeChild(self.placeholder);
			self.dragBlock.style.display = 'block';
		}

		self.isNewBlock = false;
		self.zoneLeave(zoneName);
	});
};

/**
 * Retrieve the zone for any given element
 * @param {Element} el
 */
Builder.prototype.getZone = function(el){
	var zoneName = el.getAttribute('data-zone');
	if (zoneName) return el;

	var parent = $(el).parent('[data-zone]');
	if (parent) return parent[0];
};

/**
 * Called when entering a zone
 * @param {String} zoneName
 */
Builder.prototype.zoneEnter = function(zoneName){
	var zone = this.zones[zoneName];
	if (!zone) return;

	zone.classList.add('dragover');

	if (indexOf(zone.childNodes, this.placeholder) == -1){
		zone.appendChild(this.placeholder);
	}
};

/**
 * Called when leaving a zone
 * @param {String} zoneName
 */
Builder.prototype.zoneLeave = function(zoneName){
	var zone = this.zones[zoneName];
	if (!zone) return;

	zone.classList.remove('dragover');

	if (this.isNewBlock && zone == this.placeholder.parentNode){
		zone.removeChild(this.placeholder);
	}
};

/**
 * Set back initial values when a drag ends
 *
 * If an existing block was being dragged, make sure it ends up somewhere
 */
Builder.prototype.dragEnd = function(){
	if (!this.targetZone && !this.isNewBlock){
		var zone = this.placeholder.parentNode;
		zone.insertBefore(this.dragBlock, this.placeholder);
		this.placeholder.style.height = '';
		zone.removeChild(this.placeholder);
		this.dragBlock.style.display = 'block';
	}

	this.dragging = false;
	this.isNewBlock = false;
	this.targetZone = undefined;
};

/**
 * Add block of given type to given zone
 * @param {String} zoneName
 * @param {String} type
 * @param {String} style
 * @param {Object} data
 */
Builder.prototype.addBlock = function(zoneName, type, style, data){
	if (!blocks[type]){
		return console.error('Undefined block type `%s`', type);
	}

	var block = new blocks[type](style, data),
		zone = this.zones[zoneName],
		blockId = this.blockId++;

	block.wrapper.setAttribute('data-block-id', blockId);

	this.blockIds[blockId] = block;

	if (this.placeholder.parentNode){
		zone.insertBefore(block.wrapper, this.placeholder);
		zone.removeChild(this.placeholder);
		return;
	}

	zone.appendChild(block.wrapper);
};

/**
 * Set page template to use
 * @param {String} name
 */
Builder.prototype.setTemplate = function(name){
	if (!name){
		this.template.innerHTML = '<div class="builder--empty-layout">' +
			'Choose a template to start building' +
		'</div>';
		this.zones = {};

		return;
	}
	this.template.innerHTML = templates[name];
	this.cacheZones();
};

/**
 * Serialize values from all blocks in the correct zones and block order
 * @return {Object}
 */
Builder.prototype.getValues = function(){
	var self = this, blocks, i, zones = {};
	forOwn(this.zones, function(zone, zoneName){
		zones[zoneName] = [];
		blocks = zone.querySelectorAll('.block');
		for (i = 0; i < blocks.length; i++){
			zones[zoneName].push(self.blockIds[blocks[i].getAttribute('data-block-id')].getValues());
		}
	});

	return zones;
};

module.exports = Builder;
